src/python/pants/backend/project_info/peek.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import collections.abc
+import json
+import os
+from dataclasses import asdict, is_dataclass
+from enum import Enum
+from typing import Iterable, cast
+
+from pkg_resources import Requirement
+
+from pants.engine.addresses import Address, Addresses, BuildFileAddress
+from pants.engine.console import Console
+from pants.engine.fs import DigestContents, FileContent, PathGlobs
+from pants.engine.goal import Goal, GoalSubsystem, Outputting
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule
+from pants.engine.target import Target, UnexpandedTargets
+
+
+class OutputOptions(Enum):
+    RAW = "raw"
+    JSON = "json"
+
+
+class PeekSubsystem(Outputting, GoalSubsystem):
+    """Display BUILD file info to the console.
+
+    In its most basic form, `peek` just prints the contents of a BUILD file. It can also display
+    multiple BUILD files, or render normalized target metadata as JSON for consumption by other
+    programs.
+    """
+
+    name = "peek"
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--output",
+            type=OutputOptions,
+            default=OutputOptions.JSON,
+            help="which output style peek should use",
+        )
+        register(
+            "--exclude-defaults",
+            type=bool,
+            default=False,
+            help="whether to exclude values matching default field values from normalized output",
+        )
+
+    @property
+    def output_type(self) -> OutputOptions:
+        """Get the output type from options.
+
+        Must be renamed here because `output` conflicts with `Outputting` class.
+        """
+        return cast(OutputOptions, self.options.output)
+
+    @property
+    def exclude_defaults(self) -> bool:
+        return cast(bool, self.options.exclude_defaults)
+
+
+class Peek(Goal):
+    subsystem_cls = PeekSubsystem
+
+
+def _render_raw(fcs: Iterable[FileContent]) -> str:
+    sorted_fcs = sorted(fcs, key=lambda fc: fc.path)
+    rendereds = map(_render_raw_build_file, sorted_fcs)
+    return os.linesep.join(rendereds)
+
+
+def _render_raw_build_file(fc: FileContent, encoding: str = "utf-8") -> str:
```

Comment:
Regarding encoding, we've been assuming UTF-8. We validate the locale in https://github.com/pantsbuild/pants/blob/1b47b30d7d679505f13a9ec3d38d6381284733f3/src/python/pants/bin/pants_loader.py#L45-L54

We assume UTF-8 for BUILD files by leaving off the `encoding` arg:

https://github.com/pantsbuild/pants/blob/1b47b30d7d679505f13a9ec3d38d6381284733f3/src/python/pants/engine/internals/build_files.py#L108-L111

While this certainly is not 100% safe, it's worked for the past few years as far as we can tell. I imagine we will need to revisit this if/when adding Windows support, as I think it uses UTF-16. But that's out of scope for your PR - we'll need to revisit a lot of things for that to work.